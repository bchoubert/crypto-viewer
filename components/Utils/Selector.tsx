import React, { FC } from 'react';
import {
  TouchableOpacity, View, Text, StyleSheet,
} from 'react-native';

import Colors from '../../assets/Colors';

const styles = StyleSheet.create({
  selector_container: {
    flexDirection: 'row',
    alignSelf: 'stretch',
    borderRadius: 18,
    borderTopWidth: 1,
    borderBottomWidth: 1,
    height: 35,
  },
  selector_item: {
    flex: 1,
    height: 34,
    textAlign: 'center',
    flexDirection: 'row',
    alignItems: 'center',
    borderRadius: 17,
  },
  selector_item_text: {
    color: Colors.white,
    flex: 1,
    textAlign: 'center',
  },
});

interface SelectorProps {
  items: string[];
  activeItem: string;
  setActiveItem: (item: string) => void;
  color: string;
}

const Selector: FC<SelectorProps> = ({
  items,
  activeItem,
  setActiveItem,
  color,
}) => (
  <View style={{ ...styles.selector_container, borderColor: color }}>
    {items.map((item: string) => (
      <TouchableOpacity
        key={item}
        style={{
          ...styles.selector_item,
          ...((activeItem === item) ? { backgroundColor: color } : []),
        }}
        onPress={() => setActiveItem(item)}
      >
        <Text style={{ ...styles.selector_item_text, ...((activeItem !== item) ? { color } : []) }}>
          {item}
        </Text>
      </TouchableOpacity>
    ))}
  </View>
);

export default Selector;
