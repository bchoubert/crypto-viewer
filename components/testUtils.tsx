/* eslint-disable import/no-extraneous-dependencies */
import { render, RenderAPI } from '@testing-library/react-native';
import '@testing-library/jest-dom/extend-expect';
import React, { ReactElement } from 'react';

import NavigationProvider from '../contexts/NavigationProvider';
import { SettingsContext } from '../contexts/SettingsProvider';
import ThemeProvider from '../contexts/ThemeProvider';
import TranslationProvider from '../contexts/TranslationProvider';

import { defaultSettings } from '../models/SettingsType';

jest.useFakeTimers();

export const enrichProviders = (node: ReactElement): ReactElement => (
  <SettingsContext.Provider
    value={{
      settings: defaultSettings,
      changeSettings: jest.fn(),
    }}
  >
    <TranslationProvider>
      <ThemeProvider>
        <NavigationProvider>
          {node}
        </NavigationProvider>
      </ThemeProvider>
    </TranslationProvider>
  </SettingsContext.Provider>
);

// eslint-disable-next-line max-len
export const renderNode = (node: ReactElement): RenderAPI => render(enrichProviders(node));
