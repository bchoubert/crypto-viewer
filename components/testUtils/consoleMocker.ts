/* eslint-disable no-console */
const consoleMocker = {
  consoleError: null,

  errorsToSuppress: [
    'Failed prop type',
  ],

  mockConsole: () => {
    consoleMocker.consoleError = console.error;
    console.error = (message: string) => {
      let printMessage = true;

      consoleMocker.errorsToSuppress.forEach((errorToSuppress: string) => {
        if (message.includes(errorToSuppress)) {
          printMessage = false;
        }
      });

      if (printMessage) {
        consoleMocker.consoleError?.(message);
      }
    };
  },

  restoreConsole: () => {
    if (consoleMocker.consoleError) {
      console.error = consoleMocker.consoleError;
    }
  },
};

export default consoleMocker;
