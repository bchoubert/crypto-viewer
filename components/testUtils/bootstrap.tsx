/* eslint-disable no-console */
/* eslint-disable import/no-extraneous-dependencies */
import { render, RenderAPI } from '@testing-library/react-native';
import '@testing-library/jest-dom/extend-expect';
import React, { ReactElement } from 'react';
import * as PassThoughComponents from '../PassThroughComponents';

import NavigationProvider from '../../contexts/NavigationProvider';
import { SettingsContext } from '../../contexts/SettingsProvider';
import ThemeProvider from '../../contexts/ThemeProvider';
import TranslationProvider from '../../contexts/TranslationProvider';

import { defaultSettings } from '../../models/SettingsType';
import consoleMocker from './consoleMocker';

jest.useFakeTimers();

export const enrichProviders = (node: ReactElement): ReactElement => (
  <SettingsContext.Provider
    value={{
      settings: defaultSettings,
      changeSettings: jest.fn(),
    }}
  >
    <TranslationProvider>
      <ThemeProvider>
        <NavigationProvider>
          {node}
        </NavigationProvider>
      </ThemeProvider>
    </TranslationProvider>
  </SettingsContext.Provider>
);

// eslint-disable-next-line max-len
export const renderNode = (node: ReactElement): RenderAPI => render(enrichProviders(node));

beforeAll(() => {
  consoleMocker.mockConsole();
  // eslint-disable-next-line no-extend-native
  Date.prototype.getTimezoneOffset = jest.fn(() => 0);
});

afterAll(() => {
  consoleMocker.restoreConsole();
  jest.resetAllMocks();
});

jest.spyOn(PassThoughComponents, 'Image');
(PassThoughComponents.Image as unknown as jest.Mock).mockImplementation(() => (<div>IMAGE</div>));
