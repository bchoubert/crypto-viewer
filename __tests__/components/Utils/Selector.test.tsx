import { fireEvent } from '@testing-library/react-native';
import React from 'react';

import { renderNode } from '../../../components/testUtils/bootstrap';
import Selector from '../../../components/Utils/Selector';

describe('<Selector />', () => {
  const setActiveItem = jest.fn();

  const basicItems = ['item1', 'item2'];

  it('Prints correctly', () => {
    const { toJSON } = renderNode(
      <Selector
        activeItem="item1"
        setActiveItem={setActiveItem}
        items={basicItems}
        color="#335588"
      />,
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('Basic interactions', () => {
    const node = renderNode(
      <Selector
        activeItem="item1"
        setActiveItem={setActiveItem}
        items={basicItems}
        color="#335588"
      />,
    );

    expect(setActiveItem).toHaveBeenCalledTimes(0);

    // Verify items
    basicItems.forEach((item: string) => node.getByText(item));

    // Click on other item
    fireEvent.press(node.getByText('item2'));

    expect(setActiveItem).toHaveBeenCalledTimes(1);
    expect(setActiveItem).toHaveBeenCalledWith('item2');
  });

  it('Custom renderer', () => {
    const node = renderNode(
      <Selector
        activeItem="item1"
        setActiveItem={setActiveItem}
        items={basicItems}
        color="#335588"
        // Replace "item1" by "other1"
        renderItem={(item) => `other${item[item.length - 1]}`}
      />,
    );

    expect(setActiveItem).toHaveBeenCalledTimes(1);

    // Verify items
    node.getByText('other1');
    node.getByText('other2');

    // Click on other item
    fireEvent.press(node.getByText('other2'));

    expect(setActiveItem).toHaveBeenCalledTimes(2);
    expect(setActiveItem).toHaveBeenCalledWith('item2');
  });
});
