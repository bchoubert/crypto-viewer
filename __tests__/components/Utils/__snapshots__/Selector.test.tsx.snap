// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Selector /> Prints correctly 1`] = `
<View
  style={
    Object {
      "alignSelf": "stretch",
      "borderBottomWidth": 1,
      "borderColor": "#335588",
      "borderRadius": 18,
      "borderTopWidth": 1,
      "flexDirection": "row",
      "height": 35,
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#335588",
        "borderRadius": 17,
        "flex": 1,
        "flexDirection": "row",
        "height": 34,
        "opacity": 1,
        "textAlign": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFFFFF",
          "flex": 1,
          "textAlign": "center",
        }
      }
    >
      item1
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "borderRadius": 17,
        "flex": 1,
        "flexDirection": "row",
        "height": 34,
        "opacity": 1,
        "textAlign": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#222222",
          "flex": 1,
          "textAlign": "center",
        }
      }
    >
      item2
    </Text>
  </View>
</View>
`;
